### ISSUE TITLE
Overriding Values with different record types

### ISSUE THREAD

**Issue by mrwillrobbins** [octodns/octodns#1174]
Hi,

I wanted to point out that when using split DNS views for internal/external on the same zone, if the record type is different between the two, octodns wants to create the other record type internally.  It occurs when there is an CNAME externally and with an A record internally and vice-versa.  If record types match, it works great.  For the internal view, I use both sources and set populate_should_replace to true on the internal source. 

Thank you,
Will

**Comment by mrwillrobbins** [comment]
Hi, 

I've been attempting to add the necessary logic to `octodns/provider/yaml.py` but haven't had any success so far.  The changes are needed specifically for A and CNAME record types.  I noticed that at the zone apex, issues arise with TXT and MX records.  There are scenarios where a CNAME record externally is aliased to a CDN, while the internal view points to an A record.

I appreciate that we can use `populate_should_replace` for split DNS, as it helps keep records between the views in sync when both record types should match.

Below is a sample plan output when `extview.willrobbins.com` has an A record internally but a CNAME for the same host exists externally.

(venv) [dns]$ octodns-sync --config-file=./config/internal.yaml willrobbins.com. 
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: config_file=./config/internal.yaml, (octoDNS 1.9.0)
2024-06-20T16:26:05  [140102381957376] INFO  Manager _config_executor: max_workers=1
2024-06-20T16:26:05  [140102381957376] INFO  Manager _config_include_meta: include_meta=False
2024-06-20T16:26:05  [140102381957376] INFO  Manager _config_enable_checksum: enable_checksum=False
2024-06-20T16:26:05  [140102381957376] INFO  Manager _config_auto_arpa: auto_arpa={'populate_should_replace': False, 'ttl': 1800}
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: global_processors=[]
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: global_post_processors=[]
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: provider=external (octodns.provider.yaml 1.9.0)
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: provider=internal (octodns.provider.yaml 1.9.0)
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: provider=infoblox (octoblox 0.5.0)
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: processor=acme (octodns.processor.acme 1.9.0)
2024-06-20T16:26:05  [140102381957376] INFO  Manager __init__: adding auto-arpa to processors and providers, prepending it to global_post_processors list
2024-06-20T16:26:05  [140102381957376] INFO  AutoArpa[auto-arpa] __init__: ttl=1800, populate_should_replace=False, max_auto_arpa=999, inherit_ttl=False
2024-06-20T16:26:05  [140102381957376] INFO  Manager sync: eligible_zones=['willrobbins.com.'], eligible_targets=[], dry_run=True, force=False, plan_output_fh=<stdout>, checksum=None
2024-06-20T16:26:05  [140102381957376] INFO  Manager sync:   zone=willrobbins.com.
2024-06-20T16:26:05  [140102381957376] INFO  Manager sync:     sources=['external', 'internal']
2024-06-20T16:26:05  [140102381957376] INFO  Manager sync:     processors=['auto-arpa']
2024-06-20T16:26:05  [140102381957376] INFO  Manager sync:     targets=['infoblox']
2024-06-20T16:26:05  [140102381957376] INFO  YamlProvider[external] populate:   found 15 records, exists=False
2024-06-20T16:26:05  [140102381957376] INFO  YamlProvider[internal] populate:   found 5 records, exists=False
2024-06-20T16:26:05  [140102381957376] INFO  InfoBloxProvider[infoblox] plan: desired=willrobbins.com.
2024-06-20T16:26:06  [140102381957376] WARNING InfoBloxProvider[infoblox] root NS record not supported for willrobbins.com.; ignoring it
2024-06-20T16:26:06  [140102381957376] INFO  InfoBloxProvider[infoblox] plan:   Creates=1, Updates=0, Deletes=1, Existing Records=19
2024-06-20T16:26:06  [140102381957376] INFO  Plan 
********************************************************************************
* willrobbins.com.
********************************************************************************
* infoblox (InfoBloxProvider)
*   Delete <ARecord A 900, extview.willrobbins.com., ['1.1.1.1']>
*   Create <CnameRecord CNAME 3600, extview.willrobbins.com., extview.willrobbins.com.cdn.cloudflare.net.> (external)
*   Summary: Creates=1, Updates=0, Deletes=1, Existing Records=19
********************************************************************************

I will keep trying.

Much appreciated!

Will

**Comment by ross** [comment]
Sorry, could have sworn I looked at this issue and responded, but apparently not. I do remember reading it. The "standard" way to do things when there's a bunch of shared values and a few that vary between final locations would be to use 3 `YamlProviders`

The directory structure might look like:

config/
config/internal.yaml
config/external.yaml
config/exxampled.com.yaml  # things that are shared between internal and external
config/internal
config/internal/exxampled.com.yaml. # things that are only in internal
config/external
config/external/exxampled.com.yaml. # things that are only external


internal.yaml
```yaml
---
providers:
  shared-config:
    class: octodns.provider.yaml.YamlProvider
    directory: ./config/

  internal-config:
    class: octodns.provider.yaml.YamlProvider
    directory: ./config/internal/

  internal:
    class: whatever

zones:

  '*':
    sources:
      - shared-config
      - internal-config
    target:
      - internal
```
  
external.yaml
```yaml
---
providers:
  shared-config:
    class: octodns.provider.yaml.YamlProvider
    directory: ./config/

  external-config:
    class: octodns.provider.yaml.YamlProvider
    directory: ./config/external/

  external:
    class: whatever

zones:

  '*':
    sources:
      - shared-config
      - external-config
    target:
      - external
```

**Comment by mrwillrobbins** [comment]
Hey Ross, 
S

**Comment by mrwillrobbins** [comment]
Hey Ross,
No worries.  Thank you for that solution, will proceed with that route.  I will keep working if I can use less sources, I think it would be better for operations and less confusion.  But this unblocks me from migration our our primary domain. 

Great work on this!

Will

**Comment by mrwillrobbins** [comment]
Was thinking about this, the problem is that the record types are different with the same hostname between internal and external.  The populate_should_replace works only if the record types match. I can split it up and don't use  populate_should_replace but I think that is a great feature when Ops forget to add it external but not in internal. 

config/exxampled.com.yaml # things that are shared between internal and external

**Comment by ross** [comment]
I get that, but I don't see how octoDNS/anything can tell when a record with the same name, but a different type is supposed to coexist vs when it's supposed to replace. 

At the very least i think whatever logic would be applied to get the right result would be too specific for octoDNS to implement.

**Comment by mrwillrobbins** [comment]
Very true
